alf/experience_replayers/segment_tree.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""SegmentTree."""
+
+import math
+import torch
+import torch.nn as nn
+
+import alf
+from alf.nest.utils import convert_device
+
+
+class SegmentTree(nn.Module):
+    """
+    Data structure to allow efficient calculate the summary statistics over a
+    segment of elements.
+    See https://en.wikipedia.org/wiki/Segment_tree for detail.
+
+    In this implementation, ``values[1]`` is the root. ``values[capacity: 2*capacity]``
+    are the leaves. The two children of an internal node ``values[i]`` are ``values[2*i]``
+    and ``values[2*i+1]``. And ``values[i]`` is set to ``op(values[2*i], values[2*i+1])``
+    """
```

Comment:
Should also comment that the leaves represent scalars instead of intervals in this case.

Reply:
updated.